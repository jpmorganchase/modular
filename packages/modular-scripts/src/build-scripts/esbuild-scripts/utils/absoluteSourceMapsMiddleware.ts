import * as path from 'path';
import * as fs from 'fs-extra';
import type { Request, Response } from 'express';

type SourceMap = { sources: string[] };

/* This middleware remedies this issue - https://github.com/evanw/esbuild/pull/1234,
   where paths in sources that are relative to outDir break react-error-overlay.
   This is to be used before static.
*/

export function createAbsoluteSourceMapMiddleware(outDir: string) {
  return function sourceMapMiddleware(req: Request, res: Response): void {
    const filePath = path.join(outDir, req.path);
    void fs.readJson(filePath).then((sourceMapObject: SourceMap) => {
      // Make all paths absolute
      sourceMapObject.sources = sourceMapObject.sources.map((sourcePath) =>
        path.join(outDir, sourcePath),
      );
      res.json(sourceMapObject);
    });
  };
}
