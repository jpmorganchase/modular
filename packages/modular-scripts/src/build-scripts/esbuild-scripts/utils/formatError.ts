import * as path from 'path';
import { Message } from 'esbuild';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import { codeFrameColumns } from '@babel/code-frame';

export async function formatError(
  error: Message,
  baseDir: string = process.cwd(),
): Promise<string> {
  if (error.location?.file) {
    const pathToFile = path.join(baseDir, error.location?.file);
    const source = await fs.readFile(pathToFile, {
      encoding: 'utf-8',
    });
    return `${chalk.red('Error:')}[${error.location.file}] ${error.text}
  
${codeFrameColumns(
  source,
  { start: { line: error.location.line, column: error.location.column } },
  {
    highlightCode: true,
  },
)}
    `;
  } else {
    return `${chalk.red('Error:')} ${error.text}\n`;
  }
}
