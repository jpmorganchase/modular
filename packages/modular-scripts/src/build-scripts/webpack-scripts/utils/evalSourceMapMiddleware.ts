import Server from 'webpack-dev-server';
import webpack from 'webpack';
import type {
  NextFunction,
  Request,
  Response,
  Stats,
} from 'webpack-dev-server';

function getBase64SourceMap(source: webpack.sources.Source) {
  const base64 = Buffer.from(JSON.stringify(source.map()), 'utf8').toString(
    'base64',
  );
  return `data:application/json;charset=utf-8;base64,${base64}`;
}

function getSourceById(stats: Stats, id: string) {
  const module = Array.from(stats.compilation.modules).find(
    (m) => stats.compilation.chunkGraph.getModuleId(m) === id,
  );
  return module ? module.originalSource() : null;
}

/*
 * Middleware responsible for retrieving a generated source
 * Receives a webpack internal url: "webpack-internal:///<module-id>"
 * Returns a generated source: "<source-text><sourceMappingURL><sourceURL>"
 *
 * Based on EvalSourceMapDevToolModuleTemplatePlugin.js
 */
export default function createEvalSourceMapMiddleware(server: Server) {
  return function handleWebpackInternalMiddleware(
    req: Request,
    res: Response,
    next: NextFunction,
  ) {
    if (req.url.startsWith('/__get-internal-source')) {
      const fileName = req.query.fileName as string;
      const match = fileName.match(/webpack-internal:\/\/\/(.+)/);
      const id = match ? match[1] : undefined;
      // Server.stats is private so we get around it this way
      const stats = server['stats'] as Stats | undefined;
      if (!id || !stats) {
        next();
      } else {
        const source = getSourceById(stats, id);
        if (source) {
          const base64SourceMap = getBase64SourceMap(source);
          const sourceMapURL = `//# sourceMappingURL=${base64SourceMap}`;

          const sourceURL = `//# sourceURL=webpack-internal:///${module.id}`;
          res.end(
            `${source.source() as string}\n${sourceMapURL}\n${sourceURL}`,
          );
        }
      }
    } else {
      next();
    }
  };
}
