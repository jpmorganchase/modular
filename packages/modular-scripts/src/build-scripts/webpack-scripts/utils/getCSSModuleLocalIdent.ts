import path from 'path';
import loaderUtils from 'loader-utils';

export default function getLocalIdent(
  context: { resourcePath: string; rootContext: string },
  localIdentName: unknown,
  localName: string,
  options: unknown,
) {
  // Use the filename or folder name, based on some uses the index.js / index.module.(css|scss|sass) project style
  const fileNameOrFolder = context.resourcePath.match(
    /index\.module\.(css|scss|sass)$/,
  )
    ? '[folder]'
    : '[name]';
  // Create a hash based on a the file location and class name. Will be unique across a project, and close to globally unique.
  const hash = loaderUtils.getHashDigest(
    Buffer.from(
      path.posix.relative(context.rootContext, context.resourcePath) +
        localName,
    ),
    'md5',
    'base64',
    5,
  );
  // Use loaderUtils to find the file or folder name
  const className = loaderUtils.interpolateName(
    context,
    fileNameOrFolder + '_' + localName + '__' + hash,
    options,
  );
  // Remove the .module that appears in every classname when based on the file and replace all "." with "_".
  return className.replace('.module_', '_').replace(/\./g, '_');
}
