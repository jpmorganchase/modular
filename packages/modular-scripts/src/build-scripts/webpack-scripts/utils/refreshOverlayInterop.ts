import { dismissRuntimeErrors, reportRuntimeError } from 'react-error-overlay';

module.exports = {
  clearRuntimeErrors: dismissRuntimeErrors,
  handleRuntimeError: reportRuntimeError,
};
