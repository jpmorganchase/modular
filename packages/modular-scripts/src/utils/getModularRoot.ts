import * as path from 'path';
import * as fs from 'fs-extra';
import findUp from 'find-up';

function isModularRoot(packageJson: { modular?: Record<string, unknown> }) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory: string) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

export default function getModularRoot(): string {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
}
