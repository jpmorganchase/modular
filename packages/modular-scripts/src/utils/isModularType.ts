import { JSONSchemaForNPMPackageJsonFiles as PackageJson } from '@schemastore/package';
import * as path from 'path';
import * as fs from 'fs-extra';

export const packageTypes: PackageType[] = ['app', 'view', 'package'];

export const ModularTypes: ModularType[] = (
  packageTypes as ModularType[]
).concat(['root']);

export type PackageType = 'app' | 'view' | 'package';

export type ModularType = PackageType | 'root';

export type ModularPackageJson = PackageJson & {
  browserslist?: Record<string, string[]>;
  modular?: {
    type: ModularType;
  };
};

export function getModularType(dir: string): ModularType | undefined {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = fs.readJsonSync(packageJsonPath) as ModularPackageJson;
    return packageJson.modular?.type || 'package';
  }
}

export default function isModularType(dir: string, type: PackageType): boolean {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = fs.readJsonSync(packageJsonPath) as ModularPackageJson;
    return packageJson.modular?.type === type;
  }
  return false;
}

export function isValidModularType(dir: string): boolean {
  return ModularTypes.includes(getModularType(dir) as ModularType);
}

export function isValidModularRootPackageJson(dir: string): boolean {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = fs.readJsonSync(packageJsonPath) as ModularPackageJson;
    return (
      packageJson.modular?.type === 'root' &&
      !!packageJson.private &&
      !!packageJson?.workspaces?.includes('packages/**')
    );
  }
  return false;
}
