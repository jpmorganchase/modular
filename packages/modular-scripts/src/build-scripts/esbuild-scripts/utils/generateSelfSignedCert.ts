import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import selfsigned from 'selfsigned';
import rimraf from 'rimraf';
import * as logger from '../../../utils/logger';

const del = promisify(rimraf);

// This code is mostly duplicated from webpack-dev-server, with some typings added.
// To provide consistent https behaviour in esbuild mode, we use the same directories to store the generated cert.
// See: https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js

function findCacheDir() {
  const cwd = process.cwd();

  let dir: string | undefined = cwd;

  for (;;) {
    try {
      if (fs.statSync(path.join(dir, 'package.json')).isFile()) break;
    } catch (e) {}

    const parent = path.dirname(dir);

    if (dir === parent) {
      dir = undefined;
      break;
    }

    dir = parent;
  }

  if (!dir) {
    return path.resolve(cwd, '.cache/webpack-dev-server');
  } else if (process.versions.pnp === '1') {
    return path.resolve(dir, '.pnp/.cache/webpack-dev-server');
  } else if (process.versions.pnp === '3') {
    return path.resolve(dir, '.yarn/.cache/webpack-dev-server');
  }

  return path.resolve(dir, 'node_modules/.cache/webpack-dev-server');
}

export async function generateSelfSignedCert() {
  const certificateDir = findCacheDir();
  const certificatePath = path.join(certificateDir, 'server.pem');
  let certificateExists = fs.existsSync(certificatePath);

  if (certificateExists) {
    const certificateTtl = 1000 * 60 * 60 * 24;
    const certificateStat = fs.statSync(certificatePath);

    const now = new Date();

    // cert is more than 30 days old, kill it with fire
    if ((Number(now) - Number(certificateStat.ctime)) / certificateTtl > 30) {
      logger.log('SSL Certificate is more than 30 days old. Removing...');

      await del(certificatePath);

      certificateExists = false;
    }
  }

  if (!certificateExists) {
    logger.log('Generating SSL Certificate...');

    const attributes = [{ name: 'commonName', value: 'localhost' }];
    const pems = selfsigned.generate(attributes, {
      algorithm: 'sha256',
      days: 30,
      keySize: 2048,
      extensions: [
        {
          name: 'basicConstraints',
          cA: true,
        },
        {
          name: 'keyUsage',
          keyCertSign: true,
          digitalSignature: true,
          nonRepudiation: true,
          keyEncipherment: true,
          dataEncipherment: true,
        },
        {
          name: 'extKeyUsage',
          serverAuth: true,
          clientAuth: true,
          codeSigning: true,
          timeStamping: true,
        },
        {
          name: 'subjectAltName',
          altNames: [
            {
              // type 2 is DNS
              type: 2,
              value: 'localhost',
            },
            {
              type: 2,
              value: 'localhost.localdomain',
            },
            {
              type: 2,
              value: 'lvh.me',
            },
            {
              type: 2,
              value: '*.lvh.me',
            },
            {
              type: 2,
              value: '[::1]',
            },
            {
              // type 7 is IP
              type: 7,
              ip: '127.0.0.1',
            },
            {
              type: 7,
              ip: 'fe80::1',
            },
          ],
        },
      ],
    });

    fs.mkdirSync(certificateDir, { recursive: true });
    fs.writeFileSync(certificatePath, pems.private + pems.cert, {
      encoding: 'utf8',
    });
  }

  const fakeCert = fs.readFileSync(certificatePath);

  logger.log(`SSL certificate: ${certificatePath}`);

  return fakeCert;
}
