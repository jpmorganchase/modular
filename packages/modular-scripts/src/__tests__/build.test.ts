import execa from 'execa';
import { promisify } from 'util';
import _rimraf from 'rimraf';
import tree from 'tree-view-for-tests';
import path from 'path';
import fs from 'fs-extra';

import getModularRoot from '../utils/getModularRoot';
import { ModularPackageJson } from '../utils/isModularType';

const rimraf = promisify(_rimraf);

const modularRoot = getModularRoot();

function modular(str: string, opts: Record<string, unknown> = {}) {
  return execa('yarnpkg', ['modular', ...str.split(' ')], {
    cwd: modularRoot,
    cleanup: true,
    ...opts,
  });
}

async function cleanup() {
  const packagesPath = path.join(getModularRoot(), 'packages');

  await rimraf(path.join(packagesPath, 'sample-async-package'));
  // run yarn so yarn.lock gets reset
  await execa('yarnpkg', ['--silent'], {
    cwd: modularRoot,
  });
}

beforeAll(async () => {
  await cleanup();
});

afterAll(async () => {
  await cleanup();
});

describe('WHEN building with preserve modules', () => {
  beforeAll(async () => {
    await modular(
      'add sample-async-package --unstable-type package --unstable-name sample-async-package',
      {
        stdio: 'inherit',
      },
    );

    const packageSrc = path.join(
      getModularRoot(),
      'packages',
      'sample-async-package',
      'src',
    );
    await fs.emptyDir(packageSrc);
    await fs.copy(
      path.join(__dirname, '__fixtures__', 'packages', 'sample-async-package'),
      packageSrc,
    );

    // build a package too, but preserve modules
    await modular('build sample-async-package --preserve-modules', {
      stdio: 'inherit',
    });
  });

  it('THEN expects the correct output package.json', async () => {
    expect(
      await fs.readJson(
        path.join(modularRoot, 'dist', 'sample-async-package', 'package.json'),
      ),
    ).toMatchInlineSnapshot(`
      Object {
        "dependencies": Object {},
        "files": Array [
          "README.md",
          "dist-cjs",
          "dist-es",
          "dist-types",
        ],
        "license": "UNLICENSED",
        "main": "dist-cjs/index.js",
        "module": "dist-es/index.js",
        "name": "sample-async-package",
        "typings": "dist-types/index.d.ts",
        "version": "1.0.0",
      }
    `);
  });

  it('THEN expects the correct output directory structure', () => {
    expect(tree(path.join(modularRoot, 'dist', 'sample-async-package')))
      .toMatchInlineSnapshot(`
      "sample-async-package
      ├─ README.md #1jv3l2q
      ├─ dist-cjs
      │  ├─ _virtual
      │  │  ├─ _rollupPluginBabelHelpers.js #8zp60f
      │  │  └─ _rollupPluginBabelHelpers.js.map #1a6wdsa
      │  ├─ index.js #sb8xfb
      │  ├─ index.js.map #r9dxe
      │  ├─ runAsync.js #66wsvj
      │  └─ runAsync.js.map #16teii4
      ├─ dist-es
      │  ├─ _virtual
      │  │  ├─ _rollupPluginBabelHelpers.js #14tvdhd
      │  │  └─ _rollupPluginBabelHelpers.js.map #4hotzf
      │  ├─ index.js #1lz39tw
      │  ├─ index.js.map #6hlu18
      │  ├─ runAsync.js #17jkvoa
      │  └─ runAsync.js.map #u7wk4r
      ├─ dist-types
      │  ├─ index.d.ts #12l2tmi
      │  └─ runAsync.d.ts #1iek7az
      └─ package.json"
    `);
  });

  it.each(['main', 'module', 'typings'])(
    'THEN validates the typings file exists: %s',
    async (key: keyof ModularPackageJson) => {
      const packageJson = (await fs.readJSON(
        path.join(modularRoot, 'dist', 'sample-async-package', 'package.json'),
      )) as ModularPackageJson;
      const value = packageJson[key] as string;
      expect(
        fs
          .statSync(
            path.join(modularRoot, 'dist', 'sample-async-package', value),
          )
          .isFile(),
      ).toEqual(true);
    },
  );
});
