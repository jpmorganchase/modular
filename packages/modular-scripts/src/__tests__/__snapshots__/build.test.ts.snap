// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WHEN building with preserve modules SHOULD create the correct index.js 1`] = `
"export { default } from './index2.js';
//# sourceMappingURL=index.js.map
"
`;

exports[`WHEN building with preserve modules SHOULD create the correct index2.js 1`] = `
"async function runInAsync() {
  const { runAsync } = await import('./runAsync.js');
  return runAsync();
}

export { runInAsync as default };
//# sourceMappingURL=index2.js.map
"
`;

exports[`WHEN building with preserve modules SHOULD create the correct runAsync.js 1`] = `
"function runAsync() {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log(\\"done\\");
      resolve();
    }, 1e3);
  });
}

export { runAsync };
//# sourceMappingURL=runAsync.js.map
"
`;
